src/main/java/org/jitsi/videobridge/Conference.java

```diff
@@ -1292,4 +1314,41 @@ public Object put(String key, Object value)
             return null;
         }
     }
+
+    private class LastNEndpoints
+    {
+        /**
+         * The list of endpoints ordered by speech activity and video activity (that is, endpoints with video enabled
+         * are first in the list).
+         */
+        @NotNull
+        private List<String> lastNEndpointIds = new LinkedList<>();
+
+        private void update()
+        {
+            update(false);
+        }
+
+        private void update(boolean force)
+        {
+            List<AbstractEndpoint> endpointsBySpeechActivity = speechActivity.getEndpoints();
+
+            Map<Boolean, List<AbstractEndpoint>> bySendingVideo
+                    = endpointsBySpeechActivity.stream()
+                        .collect(Collectors.groupingBy(AbstractEndpoint::isSendingVideo));
+
+            List<AbstractEndpoint> lastNEndpoints = new LinkedList<>();
+            lastNEndpoints.addAll(bySendingVideo.getOrDefault(true, Collections.emptyList()));
```

Comment:
I played with this code a bit and came up with a couple ideas.  Both are in Kotlin (I'd be all for changing `LastNEndpoints` to Kotlin, even if it meant it could no longer be a private inner class).

Option 1:
```
val (epsWithVideo, epsWithoutVideo) = endpointsBySpeechActivity.partition(AbstractEndpoint::isSendingVideo)
val lastNEndpointIds = (epsSending + epsNotSending).map(AbstractEndpoint::id)
```

Option 2: use a comparator
```
val lastNRankComparator = object : Comparator<Pair<Int, AbstractEndpoint>> {
    override fun compare(o1: Pair<Int, AbstractEndpoint>, o2: Pair<Int, AbstractEndpoint>): Int {
        // If both endpoints are sending/not sending video, compare by their original speaker rank
        if (o1.second.sendingVideo == o2.second.sendingVideo) {
            return o1.first.compareTo(o2.first)
        } else if (o1.second.sendingVideo) {
            return -1
        }
        return 1
    }
}
```
And then the code to sort:
```
val result = endpointsBySpeechActivity.mapIndexed { index, entry ->
    // Create an intermediary pair to store the original speaker rank
    Pair(index, entry)
}.sortedWith(lastNRankComparator)
```

Reply:
I agree Option 1 looks nicer (though that would be offset by having to pull it out of Conference), but I don't want to re-write and test it just because of that. Do you see a problem with this code?

Reply:
> Do you see a problem with this code?

Just that I think it's doing something simple but I find it hard to follow.  And I think being able to test it separately would be a good thing.

Reply:
I'll add tests in a separate PR, and I'll make sure to simplify the code. Can we move forward and ship this?

Reply:
Ok